### PR TITLE
Move path for iterate

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/dataflow/move_paths/builder.rs [hunk 1]
```diff
@@ -151,11 +153,13 @@ impl<'b, 'a, 'gcx, 'tcx> Gatherer<'b, 'a, 'gcx, 'tcx> {
                             place.clone()
```
Comment (oli-obk) [inline src/librustc_mir/dataflow/move_paths/builder.rs]: This `place` is wrong. It should be the one at the current projection level, but it's always the outermost one. I think you can reconstruct the place by `Place::Projection(PlaceProjection { base: proj.base, elem: proj.elem.clone() })`

File: src/librustc_mir/dataflow/move_paths/builder.rs [hunk 2]
```diff
@@ -91,86 +91,93 @@ impl<'b, 'a, 'gcx, 'tcx> Gatherer<'b, 'a, 'gcx, 'tcx> {
...
-            Place::Base(PlaceBase::Static(..)) => {
-                Err(MoveError::cannot_move_out_of(self.loc, Static))
-            }
-            Place::Projection(ref proj) => {
-                self.move_path_for_projection(place, proj)
+        place.iterate(|place_base, place_projection| {
+            let mut base = match place_base {
+                PlaceBase::Local(local) => self.builder.data.rev_lookup.locals[*local],
+                PlaceBase::Static(..) => {
+                    return Err(MoveError::cannot_move_out_of(self.loc, Static));
```
Comment (matthewjasper) [inline src/librustc_mir/dataflow/move_paths/builder.rs]: This is just returning from the closure, not the enclosing method. You'll have to create a `try_fold` like interface for this to work.

File: src/librustc_mir/dataflow/move_paths/builder.rs [hunk 3]
```diff
@@ -95,82 +95,93 @@ impl<'b, 'a, 'gcx, 'tcx> Gatherer<'b, 'a, 'gcx, 'tcx> {
...
+            for proj in place_projection {
+                let mir = self.builder.mir;
+                let tcx = self.builder.tcx;
+                let place_ty = proj.base.ty(mir, tcx).ty;
+                match place_ty.sty {
+                    ty::Ref(..) | ty::RawPtr(..) =>
+                        return Err(MoveError::cannot_move_out_of(
+                            self.loc,
+                            BorrowedContent {
+                                target_place: Place::Projection(Box::new(Projection {
```
Comment (oli-obk) [inline src/librustc_mir/dataflow/move_paths/builder.rs]: this struct constructor can just be `proj.clone()`

File: src/librustc_mir/dataflow/move_paths/builder.rs [hunk 4]
```diff
@@ -95,82 +95,93 @@ impl<'b, 'a, 'gcx, 'tcx> Gatherer<'b, 'a, 'gcx, 'tcx> {
...
+                    .entry((base, proj.elem.lift()))
+                {
+                    Entry::Occupied(ent) => *ent.get(),
+                    Entry::Vacant(ent) => {
+                        let path = MoveDataBuilder::new_move_path(
+                            &mut self.builder.data.move_paths,
+                            &mut self.builder.data.path_map,
+                            &mut self.builder.data.init_path_map,
+                            Some(base),
+                            Place::Projection(Box::new(Projection {
```
Comment (oli-obk) [inline src/librustc_mir/dataflow/move_paths/builder.rs]: same here

**Pull request merged**